### ISSUE TITLE
REPL defaults to limiting output for new types even though limit_output(STDOUT) is false

### ISSUE THREAD

**Issue by jrevels** [JuliaLang/julia#16026]
This is possibly related to #15926:

```julia
julia> limit_output(STDOUT)
false

julia> type Foo x end

julia> Base.show(io::IO, foo::Foo) = limit_output(io) ? print(io, ":p") : print(io, "Hi I'm Foo($(foo.x))!")

julia> Foo(1) # would expect this to match show(STDOUT, Foo(1))
:p

julia> show(STDOUT, Foo(1))
Hi I'm Foo(1)!
```

(edited to remove an example that used the old way of doing things)

**Comment by nalimilan** [comment]
Is this with git master? AFAIK, the new way of printing a compact representation is to check `limit_output(io)` inside `show`, and no longer define `showcompact`.

**Comment by jrevels** [comment]
Yep, I should've clarified this was v0.5. Thanks for the tip, I must've missed the discussion somewhere.

**Comment by nalimilan** [comment]
Though I didn't meant this should necessarily be closed. Maybe something could be improve here.

**Comment by jrevels** [comment]
I'll reopen it then. It appears that limiting output in the REPL is indeed by default on master (at least as of a couple days ago, when I last pulled), which doesn't seem right to me:

```julia
julia> limit_output(STDOUT)
false

julia> type Foo x end

julia> Base.show(io::IO, foo::Foo) = limit_output(io) ? print(io, ":p") : print(io, "Hi I'm Foo($(foo.x))!")

julia> Foo(1)
:p

julia> show(STDOUT, Foo(1))
Hi I'm Foo(1)!
```

**Comment by JeffBezanson** [comment]
I think this is related to the half-done state of #14052. In replutil.jl, there are `writemime` definitions for a few types (including Array) that just call `show` with `limit_output` set. writemime should be merged with show, and the REPL should be using an IOContext with limiting enabled.

**Comment by JeffBezanson** [comment]
Conceptually fixed by #16354. However, the behavior is still as described in the OP here, because the REPL does not write directly to STDOUT. It has its own io stream that wraps STDOUT with settings (which could become configurable) for interactive display. This seems OK to me. I like the idea that the REPL does not change STDOUT.
